Doc/library/typing.rst

```diff
@@ -1932,6 +1963,28 @@ Functions and decorators
    runtime we intentionally don't check anything (we want this
    to be as fast as possible).
 
+.. function:: assert_never(arg, /)
+
+   Statically assert that a line of code is unreachable.
+
+   Example::
+
+       def int_or_str(arg: int | str) -> None:
+           match arg:
+               case int():
+                   print("It's an int")
+               case str():
+                   print("It's a str")
+               case _:
+                   assert_never(arg)
```

Comment:
I am not sure about the `match` semantics: how `arg` will be narrowed, but maybe we can recommend `case _ as unreachable: assert_never(unreachable)`?

Reply:
Not sure why that would be better.

Reply:
For a more complex `match` expression, I _do_ think the `case _ as unreachable: assert_never(unreachable)` phrasing would work better. Consider:

```
match type(foo):
    case type(int)():
        ...
    case type(str)():
        ...
    case _ as unreachable:
        assert_never(unreachable)
```

Reply:
Done

Reply:
Is there a reason it's:

```python
    case _ as unreachable:
        assert_never(unreachable)
```

Instead of the slightly shorter:

```python
    case unreachable:
        assert_never(unreachable)
```

?

Reply:
No strong reason, but I feel like this makes it slightly more clear that it's a fallback case.

Reply:
Agreed that the `case _ as unreachable` phrasing is more clearly intended as a catch-all than just `case unreachable`.

Reply:
I'm not sure I disagree. It's important to teach/remind people that an unqualified capture like `case unreachable:` must (if it occurs) always be the last case. (In fact the parser warns if anything follows.)